### PR TITLE
router page made into login page

### PR DESCRIPTION
**Pull request by taylormoineau**
I commented out most of the router page and converted it temporarly to nearly a copy of the login page.

It should allow you to login, then have a useEffect which authenticates your cookie, and makes a simple state change:

currentUser = nobody
to
currentUser = whatever their id is

### CODE REVIEW

File: src/router.js [hunk 1]
```diff
@@ -1,69 +1,145 @@
...
+
+  const loginUser = async e => {
+    e.preventDefault();
+    await sendJson('login', {
+      email: emailLoginState,
+      password: passwordLoginState
+    });
+  };
+
+  useEffect(() => {
```
Comment (taylormoineau) [inline src/router.js]: Either there is a problem here in displaying the data

File: server/checkAuth.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+export const checkAuth = async (req, res) => {
+  //   if (!req.cookies.authCookie) return res.status(401).json('Not authorized!');
+  const [id, hash] = req.cookies.authCookie.split('|');
+  if (hash && (await bcrypt.compare(process.env.SECRET + id, hash))) {
+    const user = (await query(`SELECT * FROM "users" WHERE id=$1`, [id]))[0];
+    if (user) {
+      req.user = user;
+      console.log(`${id} checkAuth authenticated... so wheres the data?`);
+      return user;
```
Comment (taylormoineau) [inline server/checkAuth.js]: Or maybe theres a problem here in returning the data

Comment (andrewmaxwell) [inline server/checkAuth.js]: In order to send the data to the client, you need to do `return res.json(user)`

File: server/checkAuth.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    } else {
+      console.log(`${id} is not allowed in here!`);
+      return false;
```
Comment (andrewmaxwell) [inline server/checkAuth.js]: remove this line.

File: src/router.js [hunk 4]
```diff
@@ -1,69 +1,145 @@
...
+      setCurrentUser(result.data);
+    };
+
+    fetchResult();
+  }, []);
+
+  return (
+    <div className="container">
+      <h2>Hello, {currentUser}</h2>
+      <form onSubmit={loginUser}>
```
Comment (andrewmaxwell) [inline src/router.js]: Instead of copying the login page here, you should be able to import `LoginPage` and do `<LoginPage />` here.

**Comment by andrewmaxwell** [comment]
Ok, here's what I would do:

Change `authCookie` to just add the user info to the request if they have a valid cookie. If they don't have a cookie, do nothing (just call `next()`). If they have a cookie but it's invalid, clear it, but still call `next()`. If they have a cookie and it's valid, add the user record to the request so an endpoint can use it. But also call `next()`. Here's one way to do that:
```
export const authCookie = async (req, res, next) => {
  if (req.cookies.authCookie) {
    const [id, hash] = req.cookies.authCookie.split('|');
    if (hash && (await bcrypt.compare(process.env.SECRET + id, hash))) {
      const [user] = await query(`SELECT * FROM "users" WHERE id=$1`, [id]);
      if (user) {
        console.log(`${id} is authenticated.. yay!`);
        req.user = user;
      }
    } else {
      res
        .clearCookie('authCookie')
        .status(401)
        .json(`gtfo ${id}`);
    }
  }
  next();
};
```

In the same file, I'd add a function to make sure there's a user attached to a request. If there is, that mean's they're logged in. Otherwise, they need to gtfo.
```
export const loggedInOnly = (req, res, next) => {
  if (req.user) next();
  else res.status(401).send('Unauthorized');
};
```

Over in server/index.js, move `app.use(authCookie);` to be before any of the endpoints (remember, endpoints are the ones with urls, and they always send something back to the client. Middlewares always either modify the request/response and then call `next()` or send back an error and do not call `next()`.)

Where `app.use(authCookie);` used to be, put `app.use(loggedInOnly);` It now serves the same function as `authCookie` used to.

We also need to add an endpoint that returns the logged in user if they are logged in. So we'll add another unprotected endpoint: `app.use('/checkAuth', (req, res) => res.json(req.user));`. If they aren't logged in, it will send back undefined. Since this middleware comes after `app.use(authCookie);`, `req.user` will contain the user's data if they are logged in.

Now for the UI side. If you revert your changes to `router.js`, we won't need to change much. First, we want to have the page check to see if the user is logged in and store their data in state if they are. So add this to the top of DootRouter:
```
const [currentUser, setCurrentUser] = useState('nobody');
useEffect(() => {
  loadJson('checkAuth').then(setCurrentUser);
}, []);
```
Then, somewhere in the navbar, you can put ``{currentUser && `Logged in as ${currentUser.email}`}`` and it should show that if they are logged in.

We still need to add a Log Out button and make the Login and Register pages call `setCurrentUser` when the user logs in or registers. We can talk about how to do that when you're ready.

File: server/logout.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export const logout = (req, res) => {
+  res.clearCookie('authCookie');
+  return res.status(200).redirect('/loginPage');
```
Comment (andrewmaxwell) [inline server/logout.js]: We can't control the client from the server. We'll need to have the login form do `setCurrentUser(await loadJson('checkAuth'));` after the user has successfully logged in.

File: server/logout.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+
+  //for some reason I can't get it to work without the redirect
```
Comment (andrewmaxwell) [inline server/logout.js]: You need to send it. Our convention is to send json, so like `.json('logged out')`

File: src/router.js [hunk 7]
```diff
@@ -8,6 +8,13 @@ import {App} from './App';
 export const DootRouter = () => {
   const [currentUser, setCurrentUser] = useState('nobody');
 
+  const logout = async e => {
+    e.preventDefault();
+    sendJson('logout', {
```
Comment (andrewmaxwell) [inline src/router.js]: This can be `loadJson('logout')`

File: src/router.js [hunk 8]
```diff
@@ -51,7 +58,17 @@ export const DootRouter = () => {
                   {currentUser && `Logged in as ${currentUser.email}`}
                 </span>
               </li>
-              <li></li>
+              <li className="nav-item">
+                <form className="form-inline">
```
Comment (andrewmaxwell) [inline src/router.js]: It doesn't need to be in a form. The only time I use forms is when I want to do `onSubmit`

File: src/router.js [hunk 9]
```diff
@@ -8,6 +8,13 @@ import {App} from './App';
 export const DootRouter = () => {
   const [currentUser, setCurrentUser] = useState('nobody');
 
+  const logout = async e => {
+    e.preventDefault();
```
Comment (andrewmaxwell) [inline src/router.js]: Don't need to prevent default.

**Comment by andrewmaxwell** [comment]
I can't test it at work, they block connections to postgres on AWS. But I think we need to replace our current `useEffect` in router.js with this:
```
const checkAuth = async () => {
  setCurrentUser(await loadJson('checkAuth'));
};
useEffect(() => {
  checkAuth();
}, []);
```

Then replace `<Home />` with `<Home currentUser={currentUser} checkAuth={checkAuth} />`

Then replace the Home function at the bottom with this:
```
const Home = ({currentUser, checkAuth}) =>
  currentUser ? (
    <LoginPage checkAuth={checkAuth} />
  ) : (
    `Welcome, ${currentUser.email}`
  );
```
Notice that we're passing `checkAuth` into `Home` as well as `LoginPage`. By passing this function through, we will be able to invoke it when we're ready.

With Home containing the Login form, we can delete this from DootRouter:
```
<Route path="/loginPage">
  <LoginPage />
</Route>
```

Now in loginPage.js, we can put `{checkAuth}` into the arguments and at the bottom of the `loginUser` function, we can do `await checkAuth();`. 

I can't test it at work, but hopefully it will let you log in from the home page if you're not logged in and it will show "Welcome, <email>" when you're logged in.

File: src/router.js [hunk 10]
```diff
@@ -1,69 +1,128 @@
...
+                </Link>
+              </li>
+              <li className="nav-item">
+                <span className="navbar-text">
+                  {currentUser && `Logged in as ${currentUser.email}`}
+                </span>
+              </li>
+              <li className="nav-item">
+                <form className="form-inline">
+                  <button
```
Comment (andrewmaxwell) [inline src/router.js]: Make sure this button only appears up when someone is logged in.

Comment (taylormoineau) [reply src/router.js]: Thats what I'm trying to do right now..

File: src/router.js [hunk 11]
```diff
@@ -1,69 +1,128 @@
...
 import {BrowserRouter as Router, Switch, Route, Link} from 'react-router-dom';
-import App from './App';
-import {Register} from './Register';
+import {sendJson, loadJson} from './sendJson';
 import {LoginPage} from './loginPage';
+import {Register} from './Register';
+import {App} from './App';
+
+export const DootRouter = () => {
+  const [currentUser, setCurrentUser] = useState('nobody');
```
Comment (andrewmaxwell) [inline src/router.js]: remove `nobody`. The default should be undefined.

**Pull request merged**